### PR TITLE
v: add selector option unwrapping on if `!= none`

### LINKED ISSUE #22892: cgen error - unwrapping an option field in a struct doesn't seem to unwrap it correctly
### Describe the bug

Unwrapping an option field in a struct doesn't seem to be correctly unwrapped.

### Reproduction Steps

Code: https://play.vlang.io/p/5b48070739

```v
struct LayoutTree {
    mut:
        root ?usize
}

fn LayoutTree.new() LayoutTree {
    return LayoutTree {
        root: ?usize(none)
    }
}

fn test1() {
	root := usize(1)
	mut list := []usize {}
	list << root
	println("list: ${list}")
}

fn test2() {
	mut tree := LayoutTree.new()
	mut list := []usize {}
	tree.root = ?usize(1)
	println("tree.root: ${tree.root}")
	if tree.root != none {
		println("is not none: ${tree.root}")
    	list << tree.root
		println("parents ${list}")
	}
}

fn test3() {
	mut tree := LayoutTree.new()
	mut list := []usize {}
	tree.root = ?usize(1)
	// extract root in an other variable before unwrapping
    root := tree.root
	println("tree.root: ${root}")
	if root != none {
		println("is not none: ${root}")
    	list << root
		println("parents ${list}")
	}
}

fn main() {
	test1() // pass
	test2() // fail
	test3() // pass
}
```

### Expected Behavior

Unwrapping an option field in a struct with `if` should not ask to unwrap it again

### Current Behavior

Output:

```
code.v:26:11: error: unwrapped Option cannot be used in an infix expression
   24 |     if tree.root != none {
   25 |         println("is not none: ${tree.root}")
   26 |         list << tree.root
      |              ~~
   27 |         println("parents ${list}")
   28 |     }
Exited with error status 1
```

### Possible Solution

Extracting the struct option field in a variable, then unwrap it with a `if`.

```v
struct LayoutTree {
    mut:
        root ?usize
}

fn LayoutTree.new() LayoutTree {
    return LayoutTree {
        root: ?usize(none)
    }
}

fn main() {
	mut tree := LayoutTree.new()
	mut list := []usize {}
	tree.root = ?usize(1)
	// extract root in an other variable before unwrapping
    root := tree.root
	println("tree.root: ${root}")
	if root != none {
		println("is not none: ${root}")
    	list << root
		println("parents ${list}")
	}
}
```

### Additional Information/Context

_No response_

### V version

V 0.4.8 f1333be.5bba92a

### Environment details (OS name and version, etc.)

```
V full version: V 0.4.8 f1333be.5bba92a
OS: linux, Debian GNU/Linux 12 (bookworm) (VM)
Processor: 2 cpus, 64bit, little endian, Intel(R) Xeon(R) CPU E5-2680 v4 @ 2.40GHz

getwd: /home/admin/playground
vexe: /home/admin/v/v
vexe mtime: 2024-11-17 18:25:30

vroot: OK, value: /home/admin/v
VMODULES: OK, value: .vmodules
VTMP: OK, value: /tmp/v_0

Git version: git version 2.39.5
Git vroot status: Error: fatal: detected dubious ownership in repository at '/home/admin/v'
To add an exception for this directory, call:

	git config --global --add safe.directory /home/admin/v
.git/config present: true

CC version: cc (Debian 12.2.0-14) 12.2.0
emcc version: N/A
thirdparty/tcc status: Error: fatal: detected dubious ownership in repository at '/home/admin/v/thirdparty/tcc'
To add an exception for this directory, call:

	git config --global --add safe.directory /home/admin/v/thirdparty/tcc
 Error: fatal: detected dubious ownership in repository at '/home/admin/v/thirdparty/tcc'
To add an exception for this directory, call:

	git config --global --add safe.directory /home/admin/v/thirdparty/tcc
```

> [!NOTE]
> You can use the 👍 reaction to increase the issue's priority for developers.
>
> Please note that only the 👍 reaction to the issue itself counts as a vote.
> Other reactions and those to comments will not be taken into account.

<sub><a href="https://huly.app/guest/vlang-66f40c4d-a476b54c67-771fdd?token=eyJ0eXAiOiJKV1QiLCJhbGciOiJIUzI1NiJ9.eyJsaW5rSWQiOiI2NzNhMzdjMmU3Y2M1YTc4NTQyZDkxNDMiLCJndWVzdCI6InRydWUiLCJlbWFpbCI6IiNndWVzdEBoYy5lbmdpbmVlcmluZyIsIndvcmtzcGFjZSI6InctYWxleGFuZGVyLXZsYW5nLTY2ZjQwYzRkLWE0NzZiNTRjNjctNzcxZmRjIn0.yG-lM7mA36j62K2qaQ8GqU4WtqliyM5HEoau8Fudwxw">Huly&reg;: <b>V_0.6-21335</b></a></sub>

### PR DESCRIPTION
**Pull request by felipensp**
Fix #22892

<sub><a href="https://huly.app/guest/vlang-66f40c4d-a476b54c67-771fdd?token=eyJ0eXAiOiJKV1QiLCJhbGciOiJIUzI1NiJ9.eyJsaW5rSWQiOiI2NzNhNTBlM2U3Y2M1YTc4NTQyZDk1OGYiLCJndWVzdCI6InRydWUiLCJlbWFpbCI6IiNndWVzdEBoYy5lbmdpbmVlcmluZyIsIndvcmtzcGFjZSI6InctYWxleGFuZGVyLXZsYW5nLTY2ZjQwYzRkLWE0NzZiNTRjNjctNzcxZmRjIn0.9sxQVkpZg4nCDnHbLj5ftvGGPh3yprk_jHKesUnXXio">Huly&reg;: <b>V_0.6-21338</b></a></sub>

### CODE REVIEW

File: vlib/v/checker/tests/option_ptr_without_unwrapp_err.out [hunk 1]
```diff
@@ -1,6 +1,6 @@
 vlib/v/checker/tests/option_ptr_without_unwrapp_err.vv:7:13: error: cannot use `?&Node` as `Node`, it must be unwrapped first in argument 1 to `set_trace`
     5 | fn set_trace(n Node) {
-    6 |     if n.parent != none {
+    6 |     if n.parent == none {
     7 |         set_trace(n.parent)
       |                   ~~~~~~~~
```
Comment (spytheman) [inline vlib/v/checker/tests/option_ptr_without_unwrapp_err.out:5]: That is a now a wrong error- if the condition is `if n.parent == none {`, then `n.parent` is sure to be `none`, and it is already unwrapped.

File: vlib/v/checker/tests/assign_type_mismatch_with_generics_err.out [hunk 2]
```diff
@@ -1,7 +1,7 @@
 vlib/v/checker/tests/assign_type_mismatch_with_generics_err.vv:13:9: error: cannot assign to `b`: expected `bool`, not `fn (Bar) bool`
    11 |     mut b := false
    12 |     if f.f != none {
-   13 |         b = f.f or { panic(err) }
+   13 |         b = f.f
       |               ^
    14 |     } else {
-   15 |         b = true
+   15 |         assert false
```
Comment (spytheman) [inline vlib/v/checker/tests/assign_type_mismatch_with_generics_err.out:7]: Why did this need to be changed - it does not seem relevant to the current PR?

File: vlib/v/checker/tests/option_ptr_without_unwrapp_err.out (hunk 1 above)
Comment (spytheman) [reply vlib/v/checker/tests/option_ptr_without_unwrapp_err.out:5]: Imho restore that .vv/.out pair as it was, do not invert the condition here - it worked well before.

File: vlib/v/checker/tests/assign_type_mismatch_with_generics_err.out (hunk 2 above)
Comment (spytheman) [reply vlib/v/checker/tests/assign_type_mismatch_with_generics_err.out:7]: Imho restore this .vv/.out pair.

File: vlib/v/tests/options/option_unwrap_test.v [hunk 3]
```diff
@@ -6,7 +6,7 @@ pub mut:
 
 fn set_trace(n &Node) int {
 	if n.parent != none {
-		set_trace(n.parent or { &Node{} })
+		set_trace(n.parent)
```
Comment (spytheman) [inline vlib/v/tests/options/option_unwrap_test.v:9]: nice

File: vlib/v/checker/tests/option_ptr_without_unwrapp_err.out (hunk 1 above)
Comment (felipensp) [reply vlib/v/checker/tests/option_ptr_without_unwrapp_err.out:5]: > Imho restore that .vv/.out pair as it was, do not invert the condition here - it worked well before.

But there is no error restoring it.

File: vlib/v/checker/tests/assign_type_mismatch_with_generics_err.out (hunk 2 above)
Comment (felipensp) [reply vlib/v/checker/tests/assign_type_mismatch_with_generics_err.out:7]: `b` is bool, `f.f` is `?Fn[T]`.  So makes no sense `b = f.f` there.

Comment (felipensp) [reply vlib/v/checker/tests/assign_type_mismatch_with_generics_err.out:7]: The goal is to check the check error.

Comment (felipensp) [reply vlib/v/checker/tests/assign_type_mismatch_with_generics_err.out:7]: Well... restoring it.

**Review by spytheman — APPROVED** [review]
Excellent work.
